### PR TITLE
LSP interface definitions, POC in LinterCLI

### PR DESCRIPTION
**Pull request by fredemmott**
This builds on top of #65, and only implements the `initialize` message. It also depends on https://github.com/hhvm/hh-clilib/commit/b8ac1045c4333bca3444a454d7f0297f82c2af9f

You probably want to look at this commit-at-a-time, instead of all the changes.

It includes *all* the LSP spec interfaces, not just the ones we hope to use in HHAST.

refs #25

in vscode:

```
[Info  - 4:02:28 PM] Connection to server got closed. Server will restart.
Waiting for init
Got Init
```

### CODE REVIEW

**Comment by fredemmott** [comment]
oh hello

screen shot 2018-06-15 at 10 08 57 pm

**Comment by PranayAgarwal** [comment]
I don't think the LSP spec (in JavaScript) differentiates between nullable and optional fields in an object. Since the Hack ones are all marked optional, will there be problems if it receives `field => null` from the server?

**Comment by PranayAgarwal** [comment]
Actually I may be wrong, since TypeScript does allow `name?: bool | null`

**Comment by fredemmott** [comment]
No, especially as I'm using typecoerce instead of typeassert

File: src/__Private/LintRunErrorHandler.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+namespace Facebook\HHAST\__Private;
+
+use type Facebook\CLILib\OutputInterface;
+use namespace Facebook\HHAST\Linters;
+
+/**
+ * Interface for creating custom handlers for lint errors that are found during
+ * a hhast-lint CLI run. Some examples of hander actions are generating human
```
Comment (PranayAgarwal) [inline src/__Private/LintRunErrorHandler.php]: Missed this in the last PR, but I don't think this file only applies to CLI runs?

File: src/__Private/LintRunLSPErrorHandler.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
+<?hh // strict
```
Comment (PranayAgarwal) [inline src/__Private/LintRunLSPErrorHandler.php]: File name should ideally be `LintRunErrorHandlerLSP` to be consistent with the others

Comment (PranayAgarwal) [reply src/__Private/LintRunLSPErrorHandler.php]: Or really just use some more consistency in general (e.g. `Lint` vs `Linter`)

File: src/__Private/LintRunErrorHandler.php (hunk 1 above)
Comment (fredemmott) [reply src/__Private/LintRunErrorHandler.php]: yep

**Pull request merged**